components/script/dom/document.rs

```diff
@@ -3459,6 +3462,97 @@ impl Document {
         // implement the Permissions Policy specification.
         true
     }
+
+    /// <https://w3c.github.io/IntersectionObserver/#dom-intersectionobserver-intersectionobserver>
+    pub(crate) fn add_intersection_observer(&self, intersection_observer: &IntersectionObserver) {
+        self.intersection_observers
+            .borrow_mut()
+            .push(Dom::from_ref(intersection_observer));
+    }
+
+    /// <https://w3c.github.io/IntersectionObserver/#dom-intersectionobserver-intersectionobserver>
+    pub(crate) fn remove_intersection_observer(
+        &self,
+        intersection_observer: &IntersectionObserver,
+    ) {
+        self.intersection_observers
+            .borrow_mut()
+            .retain(|observer| *observer != intersection_observer)
+    }
+
+    /// <https://w3c.github.io/IntersectionObserver/#update-intersection-observations-algo>
+    pub(crate) fn update_intersection_observer_steps(
+        &self,
+        time: CrossProcessInstant,
+        can_gc: CanGc,
+    ) {
+        // Step 1-2
+        for intersection_observer in &*self.intersection_observers.borrow() {
+            self.update_single_intersection_observer_steps(intersection_observer, time, can_gc);
+        }
+    }
+
+    /// Step 2.1-2.2 of <https://w3c.github.io/IntersectionObserver/#update-intersection-observations-algo>
+    fn update_single_intersection_observer_steps(
+        &self,
+        intersection_observer: &IntersectionObserver,
+        time: CrossProcessInstant,
+        can_gc: CanGc,
+    ) {
+        // Step 1
+        // > Let rootBounds be observer’s root intersection rectangle.
+        let root_bounds = intersection_observer.root_intersection_rectangle(self);
+
+        // Step 2
+        // > For each target in observer’s internal [[ObservationTargets]] slot,
+        // > processed in the same order that observe() was called on each target:
+        intersection_observer.update_intersection_observations_steps(
+            self,
+            time,
+            root_bounds,
+            can_gc,
+        );
+    }
+
+    /// <https://w3c.github.io/IntersectionObserver/#notify-intersection-observers-algo>
+    pub(crate) fn notify_intersection_observers(&self) {
+        // Step 1
+        // > Set document’s IntersectionObserverTaskQueued flag to false.
+        self.intersection_observer_task_queued.set(false);
+
+        // Step 2-3
+        // > 2. Let notify list be a list of all IntersectionObservers whose root is in the DOM tree of document.
+        // > 3. For each IntersectionObserver object observer in notify list, run these steps:
+        // Iterating a copy of observer stored in the document to prevent borrow error.
+        for intersection_observer in self.intersection_observers.clone().borrow().iter() {
```

Comment:
Can we avoid the cloning by making `IntersectionObserver.callback` an `Rc<RefCell<>>`, so that there's internal mutability?

Or at least we can move the cloning to inside `invoke_callback` so that we don't clone every observer.

Reply:
I am not sure, because a callback can modify the `self.intersection_observers` completely, by using `disconnect()`, `observe()`, and `unobserve()`. I should probably wrote about that in the comment too.  We could filter the observer that won't invoke any callback and copy it before any iteration I guess.

Reply:
I see, good point! Thanks!